src/cargo/util/lints.rs

```diff
@@ -0,0 +1,241 @@
+use crate::core::FeatureValue::Dep;
+use crate::core::{Edition, FeatureMap, FeatureValue, Package};
+use crate::util::interning::InternedString;
+use crate::{CargoResult, GlobalContext};
+use annotate_snippets::{Level, Renderer, Snippet};
+use cargo_util_schemas::manifest::{TomlLintLevel, TomlToolLints};
+use pathdiff::diff_paths;
+use std::collections::{BTreeMap, HashSet};
+use std::ops::Range;
+use std::path::Path;
+use toml_edit::ImDocument;
+
+pub fn get_span(
+    document: &ImDocument<String>,
+    path: &[&str],
+    get_value: bool,
+) -> Option<Range<usize>> {
+    let mut table = document.as_item().as_table_like().unwrap();
+    let mut iter = path.into_iter().peekable();
+    while let Some(key) = iter.next() {
+        let (key, item) = table.get_key_value(key).unwrap();
+        if iter.peek().is_none() {
+            return if get_value {
+                item.span()
+            } else {
+                let leaf_decor = key.dotted_decor();
+                let leaf_prefix_span = leaf_decor.prefix().and_then(|p| p.span());
+                let leaf_suffix_span = leaf_decor.suffix().and_then(|s| s.span());
+                if let (Some(leaf_prefix_span), Some(leaf_suffix_span)) =
+                    (leaf_prefix_span, leaf_suffix_span)
+                {
+                    Some(leaf_prefix_span.start..leaf_suffix_span.end)
+                } else {
+                    key.span()
+                }
+            };
+        }
+        if item.is_table_like() {
+            table = item.as_table_like().unwrap();
+        }
+        if item.is_array() && iter.peek().is_some() {
+            let array = item.as_array().unwrap();
+            let next = iter.next().unwrap();
+            return array.iter().find_map(|item| {
+                if next == &item.to_string() {
+                    item.span()
+                } else {
+                    None
+                }
+            });
+        }
+    }
+    None
+}
+
+fn manifest_path(path: &Path, gctx: &GlobalContext) -> String {
+    diff_paths(path, gctx.cwd())
+        .unwrap_or_else(|| path.to_path_buf())
+        .display()
+        .to_string()
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct LintGroup {
+    pub name: &'static str,
+    pub default_level: LintLevel,
+    pub desc: &'static str,
+    pub edition_lint_opts: Option<(Edition, LintLevel)>,
+}
+
+const RUST_2024_COMPATIBILITY: LintGroup = LintGroup {
+    name: "rust-2024-compatibility",
+    default_level: LintLevel::Allow,
+    desc: "warn about compatibility with Rust 2024",
+    edition_lint_opts: Some((Edition::Edition2024, LintLevel::Deny)),
+};
+
+#[derive(Copy, Clone, Debug)]
+pub struct Lint {
+    pub name: &'static str,
+    pub desc: &'static str,
+    pub groups: &'static [LintGroup],
+    pub default_level: LintLevel,
+    pub edition_lint_opts: Option<(Edition, LintLevel)>,
+}
+
+impl Lint {
+    pub fn level(&self, lints: &TomlToolLints, edition: Edition) -> LintLevel {
```

Comment:
One thing worth thinking hard is that how to avoid future build failure if people set something like `-D cargo::all` and we introduce new cargo lints.

A possible way is having every new lint capped max to warning in the current edition, and relax in the next.

Reply:
One way to potentially do this is to have each lint have a `version` field similar to [`Feature`](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/features.rs#L518-L518), that is the version they are stabilized in. From there, we add a "dynamic" group `cargo:new-lints` that all lints stabilized in that version are apart of.  We then add documentation saying if a user doesn't want to break CI with `-D cargo::all`, they should have the command be:
```
cargo <cmd> -- -W cargo:new-lints -D cargo:all
```

I don't think this is perfect but it would work

Reply:
I don't think rustc or clippy maintain compatibility with `all`